nats-base-client/headers.ts

```diff
@@ -0,0 +1,194 @@
+/*
+ * Copyright 2020 The NATS Authors
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Heavily inspired by Golang's https://golang.org/src/net/http/header.go
+
+import { ErrorCode, NatsError } from "./error.ts";
+
+export interface Headers {
+  get(k: string): string;
+  set(k: string, v: string): void;
+  append(k: string, v: string): void;
+  has(k: string): boolean;
+  values(k: string): string[];
+  delete(k: string): void;
+}
+
+export class MsgHeaders implements Headers {
+  static HEADER = "NATS/1.0";
+  error?: number;
+  headers: Map<string, string[]> = new Map();
+
+  constructor() {}
+
+  size(): number {
+    let count = 0;
+    for (const [_, v] of this.headers.entries()) {
+      count += v.length;
+    }
+    return count;
+  }
+
+  equals(mh: MsgHeaders) {
+    if (
+      mh && this.headers.size === mh.headers.size &&
+      this.error === mh.error
+    ) {
+      for (const [k, v] of this.headers) {
+        const a = mh.values(k);
+        if (v.length !== a?.length) {
+          return false;
+        }
+        const vv = [...v].sort();
+        const aa = [...a].sort();
+        for (let i = 0; i < vv.length; i++) {
+          if (vv[i] !== aa[i]) {
+            return false;
+          }
+        }
+        return true;
+      }
+    }
+    return false;
+  }
+
+  static decode(a: Uint8Array): MsgHeaders {
+    const mh = new MsgHeaders();
+    const s = new TextDecoder().decode(a);
+    const lines = s.split("\r\n");
+    const h = lines[0];
+    if (h !== MsgHeaders.HEADER) {
+      const str = h.replace(MsgHeaders.HEADER, "");
+      mh.error = parseInt(str, 10);
+    } else {
+      lines.slice(1).map((s) => {
+        if (s) {
+          const [k, v] = s.split(":");
+          mh.append(k, v);
+        }
+      });
+    }
+    return mh;
+  }
+
+  toString(): string {
+    if (this.headers.size === 0) {
+      return "";
+    }
+    let s = MsgHeaders.HEADER;
+    for (const [k, v] of this.headers) {
+      for (let i = 0; i < v.length; i++) {
+        s = `${s}\r\n${k}:${v[i]}`;
```

Comment:
By convention, a space is usually added after the `:`. That may be important if you plan to compare incoming header with a "toString()" version?

Reply:
Which incidentally you are doing in "encodeHeader"

Reply:
This is mostly non-sensical issue - the minute that space is allowed in the value (it is), the value next to the colon, cannot be optional, because AFAIK, it is the user value. The mime specification again is not the right guidance for this - I am OK with writing the go version that the server and client use, we can eliminate a lot of ambiguity and treat the keys and values in the headers as they should be.

Reply:
From https://tools.ietf.org/html/rfc7230#section-3.2.4:
```
A field value might be preceded and/or followed by optional
whitespace (OWS); a single SP preceding the field-value is preferred
for consistent readability by humans.  The field value does not
include any leading or trailing whitespace: OWS occurring before the
first non-whitespace octet of the field value or after the last
non-whitespace octet of the field value ought to be excluded by
parsers when extracting the field value from a header field.
```
and again, this is what you are doing in the encoding of headers: https://github.com/nats-io/nats.deno/pull/8/files#diff-1c08cf760d969e81eb7b6018e0cc6c80R209

Reply:
👍